File: hllpp_py/src/hllpp/compare.py
############################################
# Copyright (c) 2026 Shun/修海 (@shun4midx) #
# Project: HyperLogLogPlusPlus-Autocorrect #
# File Type: Python file                   #
# File: compare.py                         #
############################################

import sys

def compare_files(file1, file2, ground_truth):
    with open(file1, 'r', encoding='utf-8') as f1, \
         open(file2, 'r', encoding='utf-8') as f2, \
         open(ground_truth, 'r', encoding='utf-8') as fg:
        out1 = [line.rstrip('\n') for line in f1.readlines()]
        out2 = [line.rstrip('\n') for line in f2.readlines()]
        gold = [line.rstrip('\n') for line in fg.readlines()]

    max_len = max(len(out1), len(out2), len(gold))
    out1 += [''] * (max_len - len(out1))
    out2 += [''] * (max_len - len(out2))
    gold += [''] * (max_len - len(gold))

    correct1 = correct2 = diff = 0

    print(f"{'Idx':<5} | {'Truth':<20} | {'File 1':<20} | {'File 2':<20} | Result")
    print("-" * 95)

    for i in range(max_len):
        t, a, b = gold[i], out1[i], out2[i]

        result = ""
        if a == t:
            correct1 += 1
            result += "OK1 "
        if b == t:
            correct2 += 1
            result += "OK2 "
        elif a != b:
            diff += 1

        print(f"{i+1:<5} | {t:<20} | {a or '[empty]':<20} | {b or '[empty]':<20} | {result or 'X'}")

    print("\nSummary")
    print("-" * 95)
    print(f"Total queries            : {max_len}")
    print(f"Correct in File 1        : {correct1} ({correct1 / max_len * 100:.2f}%)")
    print(f"Correct in File 2        : {correct2} ({correct2 / max_len * 100:.2f}%)")
    print(f"Different Suggestions    : {diff}")

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python compare.py suggestions.txt typos.txt true_answers.txt")
        sys.exit(1)

    compare_files(sys.argv[1], sys.argv[2], sys.argv[3])

File: hllpp_cpp/src/src/compare.cpp
/********************************************
 * Copyright (c) 2026 Shun/修海 (@shun4midx) *
 * Project: HyperLogLogPlusPlus-Autocorrect *
 * File Type: C++ file                      *
 * File: compare.cpp                        *
 ****************************************** */

// ======== INCLUDE ======== //
#include "../include/HLLPP/compare.h"

// ======== FUNCTION IMPLEMENTATION ======== //
void compare_files(std::filesystem::path file1, std::filesystem::path file2, std::filesystem::path ground_truth) {
    // Read all three files into string vectors
    auto read_lines = [](const std::filesystem::path& p){
        std::ifstream in(p);
        if (!in) {
            throw std::runtime_error("Unable to open " + p.string());
        }

        std::vector<std::string> lines;
        std::string line;

        while (std::getline(in, line)) {
            // rstrip '\r' if present (for Windows CRLF)
            if (!line.empty() && line.back() == '\r') 
                line.pop_back();
            lines.push_back(line);
        }

        return lines;
    };

    std::vector<std::string> out1 = read_lines(file1);
    std::vector<std::string> out2 = read_lines(file2);
    std::vector<std::string> gold = read_lines(ground_truth);

    // Pad all three to the same length
    int max_len = std::max(std::max(out1.size(), out2.size()), gold.size());
    out1.resize(max_len, "");
    out2.resize(max_len, "");
    gold.resize(max_len, "");

    // Compare
    int correct1 = 0, correct2 = 0, diff = 0;

    // Header
    std::cout << std::left << std::setw(5)  << "Idx" << " | " << std::setw(20) << "Truth" << " | " << std::setw(20) << "File 1" << " | " << std::setw(20) << "File 2" << " | Result\n"
              << std::string(95, '-') << "\n";

    for (int i = 0; i < max_len; ++i) {
        auto& t = gold[i];
        auto& a = out1[i];
        auto& b = out2[i];

        std::string result;
        if (a == t) {
            ++correct1;
            result += "OK1 ";
        }

        if (b == t) {
            ++correct2;
            result += "OK2 ";
        } else if (a != b) {
            ++diff;
        }

        auto disp = [&](const std::string& s){
            return s.empty() ? "[empty]" : s;
        };

        std::cout << std::setw(5) << (i + 1) << " | " << std::setw(20) << disp(t) << " | " << std::setw(20) << disp(a) << " | " << std::setw(20) << disp(b) << " | " << (result.empty() ? "X" : result) << "\n";
    }

    // Summary
    std::cout << "\nSummary\n" << std::string(95, '-') << "\n"
              << std::fixed << std::setprecision(2)
              << "Total queries            : " << max_len << "\n"
              << "Correct in File 1        : " << correct1 << " (" << (double)correct1 / max_len * 100 << "%)\n"
              << "Correct in File 2        : " << correct2 << " (" << (double)correct2 / max_len * 100 << "%)\n"
              << "Different Suggestions    : " << diff << "\n";
}